### ISSUE TITLE
FR: bookmarks: store also the UTC timestamp

### ISSUE THREAD

**Issue by klondi** [koreader/koreader#7652]
**Does your feature request involve difficulty completing a task? Please describe.**
Currently bookmarks are stored in whichever local time zone the OS is using. Since DST info is not stored, in time zones with daylight saving times, bookmark times can't be distinguished between from those one hour earlier when stored after the clock is sent backwards one hour. (leap second information is still kept as the %S may become 60 in that case).

**Describe the solution you'd like**
Since lua does not have any easy way to convert from UTC to the local time zone, storing the information on the local time zone is very relevant for user displayed data. On the other hand, using UTC prevents ambiguity (for example if sorting bookmarks by time) and makes it easier to handle the data when exported (I noticed this as I was writing a small tool to add the highlights and the bookmark text on the PDF itself).

I think it would be very useful if the timestamp was also stored in UTC format (by using a ! on the beggining of the format string). If desired, in a different field.
```lua
function currenttime()
  t=os.time()
  return os.date("%Y-%m-%d %H:%M:%S ",t), os.date("!%Y-%m-%d %H:%M:%S ",t)
end
```

**Describe alternatives you've considered**
Lua provides no simple way of getting the timezone information of the system, storing this is impossible, other alternative solutions would be:
Add the DST information on the time string, this may be -1 when unknown which may make things complicated and require removing this information when parsing the string to display to the user.
```lua
function currenttime()
  t=os.time()
  dst=os.date("*t",t)[dst]
  if dst == nil then dst="" elseif dst then dst="1" else dst="0" end
  return os.date("%Y-%m-%d %H:%M:%S ",t)..dst
end
```
Add the timezone information to the time string as "+HH:MM". This is a bit easier to do doing something akin to:
```lua
function currenttime()
  t=os.time()
  tdiff=os.difftime(os.time(os.date("!*t",t),t)
  if tdiff < 0 then sign="-" else sign="+" end
  if tdiff < 0 then tdiff=-tdiff else tdiff=tdiff end
  #Just in case round seconds
  tdiff=(tdiff+29)//60
  mins=tdiff%60
  hours=tdiff//60
  return os.date("%Y-%m-%d %H:%M:%S ",t)..sign..hours..":"..mins
end
```

**Comment by poire-z** [comment]
To be considered with a future rewrite of our bookmarks/highlights, cf #5794.
Probably not to be fixed before this whole rewrite, to not have to migrate intermediate versions.

**Comment by NiLuJe** [comment]
It's rather hellish to do right outside of glibc.

i.e., [something like this](https://github.com/koreader/koreader-base/blob/80bcdf5bc7aa7f42135c08c31ed2ad99c358c181/ffi/rtc.lua#L320-L321) becomes [something like that](https://github.com/koreader/koreader/blob/310151fac9b516c66ce3198abbae4c337252e6a2/plugins/calibre.koplugin/search.lua#L571-L589).

TL;DR: meh.

**Comment by NiLuJe** [comment]
Okay, now that I'm not thinkng about a different piece of code, the simplest solution is just to store an Epoch instead of a human-readable broken down calendar time.

Let the calendar shenanigans to the display code.

In any case, like @poire-z said: add it to the #5794 pile ;).

**Comment by Frenzie** [comment]
> Okay, now that I'm not thinking about a different piece of code, the simplest solution is just to store an Epoch instead of a (local or not, although local is admittedly worse, agree with you there) human-readable broken down calendar time.

But will it actually be an epoch or a "fake" TZ-adjusted epoch?

**Comment by NiLuJe** [comment]
If you don't feed anything to os.time, it just calls `time(NULL)`, which returns a "true" Epoch ;).

Issue arise when you feed it a calendar-time, because is uses `mktime` to convert it to seconds, and that enforces local time ;).

**Comment by klondi** [comment]
IIF you assume the result of os.time is an epoch you can add or substract to conversions from and to unix timestamps work as you'd expect:

```lua
to_unix_ts, from_unix_ts = (function ()
  uts=os.time({year = 1970, month = 1, day = 1, hour = 0, min = 0, sec = 0})
  utcdiff=os.difftime(os.time(os.date("!*t",uts)), uts)
  function to_unix_ts(t)
    if t == nil then t=os.time() end
    return os.difftime(t, uts)+utcdiff
  end

  --Not fully portable since we have no (easy) way of reversing os.difftime
  function from_unix_ts(t)
    return uts+(t-utcdiff)
  end
  return to_unix_ts, from_unix_ts
end) ()
```
You can save a UNIX timestamp using ts=to_unix_ts(t) (it will get the current time using os.time if no arguments passed). You can then get the original os.time() output using from_unix_ts(ts). You can use this result then with os.date for formating ensuring the formated time is in local (ot if ! is the first character, UTC) time.

**Comment by NiLuJe** [comment]
os.date is not an issue, as long as you're always using true epochs.

```
> print(os.date("%x %X", os.time()))
05/09/21 19:28:49
> print(os.date("!%x %X", os.time()))
05/09/21 17:28:54
```

Where shit gets weird is when you have "local" epochs or UTC calendar time and need to massage that into something sensible (i.e., a true epoch).

**Comment by NiLuJe** [comment]
TL;DR: Always store true epochs (os.time()), and always display local time (os.date()).

All hell breaks loose when you start to store calendar time instead; like what the current code appears to be doing ;).

**Comment by NiLuJe** [comment]
Oh, indeedn, I am 100% assuming "POSIX" here, windows be damned ;).

**Comment by NiLuJe** [comment]
And we also only care about LuaJIT's implementation (althoug hthat shouldn't really be relevant here):

https://github.com/LuaJIT/LuaJIT/blob/8dc3cd6c84dfc81539604340b7884408e1c71d55/src/lib_os.c#L171-L259

**Comment by klondi** [comment]
Okay then you can avoid the Y2k38 bug...
```lua
(function ()
  local uts=os.time({year = 1970, month = 1, day = 1, hour = 0, min = 0, sec = 0})
  uts=os.time({year = 1970, month = 1, day = 1, hour = 0, min = 0, sec = os.difftime(os.time(os.date("!*t",uts)), uts)})
  function from_unix_ts(t)
    --Not portable but avoids Y2k38
    return t+uts
  end
  function to_unix_ts(t)
    if t == nil then t=os.time() end
    return os.difftime(t, uts)
  end
end) ()
```
That should ensure true UNIX timestamps in all systems (and maybe save you some headaches in the future) and allow back conversion on most.  The more portable idiom would be this, but it at least shows the 2k38 bug on my system (not sure if I'm using luajit here but in c the seconds field struct is a 32 bit int):
```lua
  function from_unix_ts(t)
    local tt = os.date("*t",uts)
    tt.sec = tt.sec + t
    return os.time(tt)
  end
```

**Comment by NiLuJe** [comment]
FWIW, YK38 is going to be an issue regardless, because those devices are unlikely to ever see a kernel & userland with sane data types ;).

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
If this is added, please allow the UNIX Epoch, if that is the type of Epoch that the information shall be stored as, to be presented to the user if they prefer that to dates. I'm actually able to parse UNIX Epoch time quicker than I am able to parse a date...
